File: src/main/scala/com/github/ldaniels528/interview/scalaprep/StringMath.scala
package com.github.ldaniels528.interview.scalaprep

object StringMath extends App {

  val number0 = "1482"
  val number1 = "12"
  println(s"$number0 + $number1 = ${add(number0, number1)}")
  println(s"$number0 * $number1 = ${multiply(number0, number1)}")

  def multiply(numberL: String, numberR: String) = {
    val products = numberR.reverse.map(_ - '0') map { digitR =>
      var carry = 0

      // generate the product
      (numberL.reverse.map(_ - '0') map { digitL =>
        val product = digitL * digitR + carry
        val digit = product % 10
        carry = product / 10
        digit
      }).mkString.reverse
    }

    products match {
      case Seq(product) => product
      case list =>
        list.tail.zipWithIndex.foldLeft(list.head) {
          case (sum, (product, index)) => add(sum, product + ("0" * (1 + index)))
        }
    }
  }

  def add(numberL: String, numberR: String): String = {
    val limit = Math.max(numberL.length, numberR.length)
    val digitsL = numberL.map(_ - '0').reverse
    val digitsR = numberR.map(_ - '0').reverse
    var carry = 0
    (for {
      n <- 0 until limit
    } yield {
      val digitL = if (n < digitsL.length) digitsL(n) else 0
      val digitR = if (n < digitsR.length) digitsR(n) else 0
      val sum = digitL + digitR + carry
      val digit = sum % 10
      carry = sum / 10
      digit
    }).mkString.reverse
  }

}

File: src/main/java/com/github/ldaniels528/interview/javaprep/StringMath.java
package com.github.ldaniels528.interview.javaprep;

import java.util.ArrayList;
import java.util.List;

public class StringMath {

    public static void main(String[] args) {
        final String number0 = "1482";
        final String number1 = "12";

        System.out.println(String.format("%s * %s = %s", number0, number1, multiply(number0, number1)));
        System.out.println(String.format("%s + %s = %s", number0, number1, add(number0, number1)));
    }

    private static String multiply(final String numberL, final String numberR) {
        final char[] caL = numberL.toCharArray();
        final char[] caR = numberR.toCharArray();
        final int limit = Math.max(caL.length, caR.length);
        final List<String> numbers = new ArrayList<>(caR.length);

        for (int i = caR.length - 1; i >= 0; i--) {
            final int digitR = caR[i] - '0';
            int carry = 0;

            final StringBuilder sb = new StringBuilder(limit + 1);
            for (int n = i + 1; n < caR.length; n++) sb.append('0');

            for (int j = caL.length - 1; j >= 0; j--) {
                final int digitL = caL[j] - '0';
                final int product = digitL * digitR + carry;
                final int digit = product % 10;
                carry = product / 10;
                sb.append(digit);
            }
            numbers.add(sb.reverse().toString());
        }

        // generate the sum of the products
        String outcome = numbers.get(0);
        for (int n = 1; n < numbers.size(); n++) {
            outcome = add(outcome, numbers.get(n));
        }
        return outcome;
    }

    private static String add(final String numberL, final String numberR) {
        final char[] caL = numberL.toCharArray();
        final char[] caR = numberR.toCharArray();
        final int limit = Math.max(caL.length, caR.length);

        final StringBuilder sb = new StringBuilder(limit + 1);
        int carry = 0;
        for (int n = 0; n < limit; n++) {
            final int digitL = caL.length - n > 0 ? caL[(caL.length - 1) - n] - '0' : 0;
            final int digitR = caR.length - n > 0 ? caR[(caR.length - 1) - n] - '0' : 0;
            final int sum = digitL + digitR + carry;
            final int digit = sum % 10;
            carry = sum / 10;
            sb.append(digit);
        }
        return sb.reverse().toString();
    }

}
